awx/sso/backends.py

```diff
@@ -383,31 +381,73 @@ def on_populate_user(sender, **kwargs):
             force_user_update = True
             logger.warning('LDAP user {} has {} > max {} characters'.format(user.username, field, max_len))
 
-    # Update organization membership based on group memberships.
     org_map = getattr(backend.settings, 'ORGANIZATION_MAP', {})
+    team_map = getattr(backend.settings, 'TEAM_MAP', {})
+
+    # Move this junk into save of the settings for performance later, there is no need to do that here
+    #    with maybe the exception of someone defining this in settings before the server is started?
+    # ==============================================================================================================
+
+    # Get all of the names of orgs in the DB and create any new org defined in LDAP that does not exist in the DB
+    existing_org_names = list(Organization.objects.all().values_list('name', flat=True))
+    for org_name in org_map.keys():
+        if org_name not in existing_org_names:
+            Organization.objects.get_or_create(name=org_name)
+            # Add the org name to the existing orgs since we created it and we may need it to build the teams below
+            existing_org_names.append(org_name)
+
+    # Do the same for teams
+    existing_team_names = list(Team.objects.all().values_list('name', flat=True))
+    for team_name, team_opts in team_map.items():
+        if 'organization' not in team_opts:
+            logger.debug("Team named {} in LDAP team map settings is invalid due to missing organization".format(team_name))
+            continue
+        # A team may have an org that we didn't previously evaluate
+        if team_opts['organization'] not in existing_org_names:
+            Organization.objects.get_or_create(name=team_opts['organization'])
+            # Append it to the list so that we don't try and create this again if its used by another team
+            existing_org_names.append(team_opts['organization'])
+        if team_name not in existing_team_names:
+            # Going a little inefficient here, we could prob get name/ids above and then set this by ID but I've got limited time r/n
+            #    and this code should only really be getting hit once if an LDAP config change adds a team with an org not built.
+            org, created = Organization.objects.get_or_create(name=team_opts['organization'])
```

Comment:
Again, the use of `get_or_create` here doesn't make any sense to me, since all the prior checks just validated that the organization _already exists_ so this should only be a get, and never a create.

Reply:
We can't technically guarantee that because there is no restriction that an organization tied to a team has to be in the org map. Consider an org map like:
```
{ "a", "b", "c"}
```
And a team map of:
```
{ "team1": { "org": "a" }, "team2": { "org": "d" } }
```
We would have never created "d" before this point.

Reply:
That would hit `if team_opts['organization'] not in existing_org_names:` in the conditional right before this. So in that scenario, we would hit line 407 and line 413, and it doesn't make sense to me that both would be `get_or_create`. In both cases, there should be no ambiguity about whether it's getting or creating. It's one or the other for each. It may create and then get.